Sync keyboard navigation in SelectionList and PopoverMenu

WojtekBoman opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details

This PR fixes focusing on SelectionList and PopoverMenu items when navigating with tab button and arrows. Previously, these buttons worked independently, now only one item on the list is focused when navigating using them.

How it works without a fix:

https://github.com/Expensify/App/assets/47774969/2d8eb991-69fb-4833-b555-e1b0725efc79



### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/36476#issuecomment-2020476802
PROPOSAL: N/A


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

1. Verify that focusing on SelectionList items works properly:
      1. Create a new workspace
      2. Open Categories page
      3. Navigate through the list of categories with tab and arrows and verify that only one element is focused.
 2. Verify that focusing on PopoverMenu items works properly:
      1. Open Workspaces section in Account Settings.
      2. Click on the three dot menu on any workspace.
      3. Navigate through the options of PopoverMenu with tab and arrows and verify that only one element is focused.

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
5. Upload an image via copy paste
6. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/47774969/1f823b43-ebc2-476b-a7f2-dbf3aaf579eb


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/47774969/fc952122-990e-44be-ad25-73d282e6148e


</details>

WojtekBoman committed:
Sync keyboard navigation in SelectionList and PopoverMenu

WojtekBoman committed:
Fix lint

WojtekBoman committed:
Run prettier and fix lint warnings

WojtekBoman committed:
Merge branch 'main' into fix/sync-arrows-and-tab-navigation

WojtekBoman committed:
Add useSyncFocus hook

WojtekBoman commented:
I'm still investigating this bug that appears on mobile browsers, I'll let you know when it's ready to check :)

WojtekBoman committed:
Merge branch 'main' into fix/sync-arrows-and-tab-navigation

WojtekBoman committed:
Fix passing ref to children of Hoverable when hover is not supported

WojtekBoman committed:
Add docs for useSyncFocus hook

Kicu commented on src/hooks/useSyncFocus/index.ts:
```diff
@@ -0,0 +1,20 @@
+
+/**
+ * Custom React hook created to handle sync of focus on an element when the user navigate through the app with keyboard.
+ * When the user navigate through an app using the arrows and then the tab button, the focus on the element and the native focus of the browser differs.
```
```suggestion
 * When the user navigates through the app using the arrows and then the tab button, the focus on the element and the native focus of the browser differs.
```

Kicu commented on src/hooks/useSyncFocus/index.ts:
```diff
@@ -0,0 +1,20 @@
+            return;
+        }
+
+        ref?.current?.focus();
```
```suggestion
        ref.current?.focus();
```

I believe the ref itself is always there if the type is `RefObject<HTMLDivElement | View>`. I understand that bugs may appear in runtime, but the code should be written according to TS imo.

Kicu commented on src/components/PopoverMenu.tsx:
```diff
@@ -193,7 +194,7 @@ function PopoverMenu({
                 {enteredSubMenuIndexes.length > 0 && renderBackButtonItem()}
                 {currentMenuItems.map((item, menuIndex) => (
-                    <MenuItem
+                    <PopoverMenuListItem
```
the file is called `PopoverMenuItem` but the component is called `PopoverMenuListItem`.
Thats probably a mistake? I think `PopoverMenuItem` sounds good

Kicu reviewed (approved):
Review after fixing the focus bug.
LGTM 👍 good job

WojtekBoman committed:
Refactor useSyncFocus, rename PopoverMenuItem to FocusableMenuItem

WojtekBoman commented:
@fedirjh I believe that now it should work, but I'm not sure about bugs on `Android:Native` I can't reproduce that, for me Tab doesn't navigate on Native platforms. Issues that occur on mobile browsers have been resolved :)

fedirjh reviewed (approved):
Looks good to me.

WojtekBoman committed:
Merge branch 'main' into fix/sync-arrows-and-tab-navigation

pecanoro commented:
We merged [this](https://github.com/Expensify/App/pull/39327) to fix some bugs with the key arrow functionality, since this is a considerable refactor of similar logic, please make sure the bug doesn't show up here.

WojtekBoman committed:
Merge branch 'main' into fix/sync-arrows-and-tab-navigation

WojtekBoman committed:
Add isFocused prop to useArrowKeyFocusManager

WojtekBoman committed:
revert

WojtekBoman committed:
Merge branch 'main' into fix/sync-arrows-and-tab-navigation

WojtekBoman committed:
Add isFocused prop to useArrowKeyFocusManager

WojtekBoman commented:
> We merged [this](https://github.com/Expensify/App/pull/39327) to fix some bugs with the key arrow functionality, since this is a considerable refactor of similar logic, please make sure the bug doesn't show up here.

In this PR I replaced ArrowKeyFocusManager wrapper with useArrowKeyFocusManager hook in BaseSelectionList, so to fix that issue I've added the `isFocused` prop to the hook and it seems to work fine :)

https://github.com/Expensify/App/assets/47774969/aebf393c-1fa0-4e15-a90c-1fd01ad6b7dc

dannymcclain reviewed (dismissed):
Looking good from a design perspective! This is a great update. `#not-an-engineer`

roryabraham commented:
conflicts

WojtekBoman committed:
Merge branch 'main' into fix/sync-arrows-and-tab-navigation

WojtekBoman commented:
> @WojtekBoman Could you please check the failing Reassure Performance Test?

Yes, I'll take a look at it. I'll let you know when it's ready :)

WojtekBoman committed:
Fix onFocusedIndexChange in BaseSelectionList

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

WojtekBoman commented:
@Kicu has found a regression related to this PR. I've already prepared a fix for that: https://github.com/Expensify/App/pull/40234. You can find more details in the description of the attached PR

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8732363773) to staging by https://github.com/roryabraham in version: 1.4.63-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

mallenexpensify commented:
Possible regression from here, can @WojtekBoman and @fedirjh , 👀 plz. 
[FAB – Blue frame appears when navigate via FAB menu using Up/Down keys](https://github.com/Expensify/App/issues/40398#top)

cristipaval commented:
Another possible regression: https://github.com/Expensify/App/issues/40410

jayeshmangwani commented on src/components/SelectionList/BaseSelectionList.tsx:
```diff
@@ -225,6 +222,17 @@ function BaseSelectionList<TItem extends ListItem>(
+            scrollToIndex(index, true);
+        },
+        isFocused,
+    });
```
When we are using the `useArrowKeyFocusManager` here, we should have passed the `disabledIndexes` props, which caused this issue: https://github.com/Expensify/App/issues/40583

jayeshmangwani commented on src/components/SelectionList/BaseListItem.tsx:
```diff
@@ -64,6 +73,7 @@ function BaseListItem<TItem extends ListItem>({
                 onMouseDown={shouldPreventDefaultFocusOnSelectRow ? (e) => e.preventDefault() : undefined}
                 nativeID={keyForList ?? ''}
                 style={pressableStyle}
+                onFocus={onFocus}
```
Coming from this issue https://github.com/Expensify/App/issues/41922, we have added the onFocus to BaseListItem in this PR. We already had the onMouseDown conditional prop passed, so the bug might have been there with the conditional onMouseDown prop but became visible after the onFocus change. When we long press the currency in CurrencySelection, it triggers the onFocus function, and we see a scroll to the long-pressed index. We have fixed this by removing the conditional onMouseDown in BaseListItem.
